dev/AppNotifications/AppNotificationUtility.cpp

```diff
@@ -331,12 +264,51 @@ winrt::guid Microsoft::Windows::AppNotifications::Helpers::RegisterComActivatorG
         THROW_IF_FAILED(hr);
     }
 
-    std::wstring notificationAppId{ RetrieveNotificationAppId() };
+    std::wstring notificationAppId{ Microsoft::Windows::AppNotifications::Helpers::RetrieveNotificationAppId() };
     RegisterAssets(notificationAppId, registeredGuid);
 
     // Remove braces around the guid string
     return winrt::guid(registeredGuid.substr(1, registeredGuid.size() - 2));
-    
+}
+
+void Microsoft::Windows::AppNotifications::Helpers::RegisterAssets(std::wstring const& appId, std::wstring const& clsid)
+{
+    wil::unique_hkey hKey;
+    // subKey: \Software\Classes\AppUserModelId\{AppGUID}
+    std::wstring subKey{ c_appIdentifierPath + appId };
+
+    THROW_IF_WIN32_ERROR(RegCreateKeyEx(
+        HKEY_CURRENT_USER,
+        subKey.c_str(),
+        0,
+        nullptr /* lpClass */,
+        REG_OPTION_NON_VOLATILE,
+        KEY_ALL_ACCESS,
+        nullptr /* lpSecurityAttributes */,
+        &hKey,
+        nullptr /* lpdwDisposition */));
+
+    // Try the following techniques to retrieve display name and icon:
+    // 1. Based on the best app shortcut, using the FrameworkUdk.
+    // 2. From the current process.
+    // 3. Set a default DisplayName, but leave empty the icon file path so Shell can set a default icon.
+    Microsoft::Windows::AppNotifications::ShellLocalization::AppNotificationAssets assets{};
+
+    if (FAILED(Microsoft::Windows::AppNotifications::ShellLocalization::RetrieveAssetsFromShortcut(assets)) &&
+        FAILED(Microsoft::Windows::AppNotifications::ShellLocalization::RetrieveAssetsFromProcess(assets)))
+    {
+        assets.displayName = GetDisplayNameBasedOnProcessName();
+    }
+
+    RegisterValue(hKey, L"DisplayName", reinterpret_cast<const BYTE*>(assets.displayName.c_str()), REG_EXPAND_SZ, assets.displayName.size() * sizeof(wchar_t));
+
+    // If IconUri is empty, then Shell will render a default icon in AppNotifications.
+    if (!assets.iconFilePath.empty())
+    {
+        RegisterValue(hKey, L"IconUri", reinterpret_cast<const BYTE*>(assets.iconFilePath.c_str()), REG_EXPAND_SZ, assets.iconFilePath.size() * sizeof(wchar_t));
```

Comment:
What would happen the second time if the App Registers but there is no icon specified?
I think we need to clear the registry of all these assets before we repopulate it.

Reply:
Otherwise we run the risk of having a stale icon from a prior registration.

Reply:
One solution is to delete the icon from registry if the iconFilePath is empty